x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/xpack/searchablesnapshots/cache/CacheFile.java

```diff
@@ -0,0 +1,275 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+package org.elasticsearch.xpack.searchablesnapshots.cache;
+
+import org.apache.lucene.store.AlreadyClosedException;
+import org.elasticsearch.action.ActionListener;
+import org.elasticsearch.common.CheckedBiConsumer;
+import org.elasticsearch.common.CheckedBiFunction;
+import org.elasticsearch.common.Nullable;
+import org.elasticsearch.common.util.concurrent.AbstractRefCounted;
+import org.elasticsearch.common.util.concurrent.ReleasableLock;
+
+import java.io.IOException;
+import java.io.UncheckedIOException;
+import java.nio.channels.FileChannel;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.nio.file.StandardOpenOption;
+import java.util.Collections;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Objects;
+import java.util.Set;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.locks.ReentrantReadWriteLock;
+
+class CacheFile {
+
+    @FunctionalInterface
+    public interface EvictionListener {
+        void onEviction(CacheFile evictedCacheFile);
+    }
+
+    static final int RANGE_SIZE = 1 << 15;
+
+    private static final StandardOpenOption[] OPEN_OPTIONS = new StandardOpenOption[]{
+        StandardOpenOption.READ, StandardOpenOption.WRITE, StandardOpenOption.CREATE, StandardOpenOption.SPARSE
+    };
+
+    private final AbstractRefCounted refCounter = new AbstractRefCounted("CacheFile") {
+        @Override
+        protected void closeInternal() {
+            CacheFile.this.finishEviction();
+        }
+    };
+
+    private final ReleasableLock evictionLock;
+    private final ReleasableLock readLock;
+
+    private final SparseFileTracker tracker;
+    private final String name;
+    private final Path file;
+
+    private volatile Set<EvictionListener> listeners;
+    private volatile FileChannel channel;
+    private volatile boolean closed;
```

Comment:
Suggest renaming this to `evicted` now.

Reply:
Makes sense, I pushed https://github.com/elastic/elasticsearch/pull/50693/commits/48fa92350a887c78dc45ed51c05d24aecdb9f570